File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrDocumentWorkspaceInfoUIExtension.cs
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrDocumentWorkspaceInfoUIExtension :
        CardUIExtension
    {
        #region Constructors

        public KrDocumentWorkspaceInfoUIExtension(IKrTypesCache typesCache)
        {
            this.typesCache = typesCache;
        }

        #endregion

        #region Fields

        private readonly IKrTypesCache typesCache;

        #endregion

        #region Private Methods

        private static void OnCardModelInitialized(object sender, CardModelInitializingEventArgs e)
        {
            StringDictionaryStorage<CardSection> sections = e.CardModel.Card.TryGetSections();
            if (sections == null
                || !sections.TryGetValue("DocumentCommonInfo", out var section))
            {
                return;
            }

            string docTypeTitle = section.RawFields.TryGet<string>("DocTypeTitle");

            if (!string.IsNullOrWhiteSpace(docTypeTitle))
            {
                e.WorkspaceInfo = docTypeTitle;
            }
        }

        #endregion

        #region Base Overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            ICardEditorModel editor;

            if (context.Model.InSpecialMode()
                && context.Model.Flags.HasNot(CardModelFlags.EditTemplate)
                && context.Model.Flags.HasNot(CardModelFlags.ViewExported)
                || (editor = context.UIContext.CardEditor) == null)
            {
                return;
            }

            editor.CardModelInitialized -= OnCardModelInitialized;

            if ((await KrComponentsHelper.GetKrComponentsAsync(context.Model.CardType.ID, this.typesCache, context.CancellationToken))
                .Has(KrComponents.DocTypes))
            {
                editor.CardModelInitialized += OnCardModelInitialized;
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/krProcess/krDocumentWorkspaceInfoUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext, CardModelFlags, CardModelInitializingEventArgs } from 'tessa/ui/cards';
import { hasNotFlag, hasFlag } from 'tessa/platform';
import { getKrComponents, KrTypesCache, KrComponents } from 'tessa/workflow';

export class KrDocumentWorkspaceInfoUIExtension extends CardUIExtension {

  public initialized(context: ICardUIExtensionContext) {
    const editor = context.uiContext.cardEditor;

    if (!editor
      || context.model.inSpecialMode
      && hasNotFlag(context.model.flags, CardModelFlags.EditTemplate)
      && hasNotFlag(context.model.flags, CardModelFlags.ViewExported)
    ) {
      return;
    }

    editor.cardModelInitialized.remove(KrDocumentWorkspaceInfoUIExtension.onCardModelInitialized);

    const usedComponents = getKrComponents(context.model.cardType.id!, null, KrTypesCache.instance);
    if (hasFlag(usedComponents, KrComponents.DocTypes)) {
      editor.cardModelInitialized.add(KrDocumentWorkspaceInfoUIExtension.onCardModelInitialized);
    }
  }

  private static async onCardModelInitialized(e: CardModelInitializingEventArgs) {
    const sections = e.cardModel.card.tryGetSections();
    if (!sections) {
      return;
    }

    const section = sections.tryGet('DocumentCommonInfo');
    if (!section) {
      return;
    }

    const docTypeTitle = section.fields.tryGet('DocTypeTitle');
    if (docTypeTitle) {
      e.workspaceInfo = docTypeTitle;
    }
  }

}